src/machine.rs

```diff
@@ -203,23 +227,21 @@ impl<'a, 'tcx> Machine<'a, 'tcx> {
 
     pub fn push_frame(&mut self,
                       def_id: DefId,
-                      mut args: Vec<TyVal<'tcx>>,
+                      args: Vec<TyVal<'tcx>>,
                       ret_val: Option<Address>,
                       ret_block: Option<BasicBlock>){
         let def = Instance::mono(self.tcx, def_id).def;
         let mir = self.load_mir(def);
 
         // Initialise locals
         let size = mir.local_decls.len();
-        let mut tmp_locals = vec![None; size];
-        for i in (0..mir.arg_count).rev() {
-            // Reverse order is an optimisation. Popping means that we
-            // don't shift the vec each time after a remove
-            let arg = args.pop().unwrap().to_bytes();
-            tmp_locals[i+1] = Some(arg); // +1 because of ret val as first index
+        let mut locals: IndexVec<Local, Vec<u8>> = IndexVec::with_capacity(size);
+        let byte_args: IndexVec<Local, Vec<u8>> = args.into_iter().map(|x| x.to_bytes()).collect();
+        locals.extend(byte_args);
+        for decl in mir.local_decls.iter().skip(mir.arg_count) {
+            let size = size_of(decl.ty);
+            locals.push(vec![0; size]);
```

Comment:
Would it be cheaper to count the number of zeros you need and then push once?

Reply:
The type of locals is `IndexVec<Local, Vec<u8>>` though, it's not a contiguous block of memory. I need to push a new `Vec` for each local.

Reply:
From a performance POV: ouch :) This will be a *lot* slower than having a single contiguous block of memory for locals: malloc and free are quite expensive, and the `Vec`s may well be allocated at different points in memory, which doesn't play well with the cache. This isn't something we should fix in this PR, but maybe for the near future? You'll want to benchmark before and after though, so it doest rest of having a benchmark of sorts!

Reply:
Agreed. You might be able to get this faster, but it depends what Rust will allow. Save it for another PR. Perhaps raise an issue?

Reply:
Wait, I see the issue now. What Laurie said: ideally you want one vector, not many. Is it hard?

Reply:
Ideally we would translate "lookup var X" to "lookup var at offset Y". That may or may not be difficult given that we're currently interpreting MIR directly.

Reply:
It starts to get a little fiddly when dealing with field accesses within aggregate types, but for the most part I don't think that it would be too hard.